shared/prometheus/content_negotiation.go

```diff
@@ -0,0 +1,59 @@
+package prometheus
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/golang/gddo/httputil"
+)
+
+const (
+	contentTypePlainText = "text/plain"
+	contentTypeJSON      = "application/json"
+)
+
+// generatedResponse is a container for response output.
+type generatedResponse struct {
+	// Err is protocol error, if any.
+	Err string `json:"error"`
+
+	// Data is response output, if any.
+	Data interface{} `json:"data"`
+}
+
+// negotiateContentType parses "Accept:" header and returns preferred content type string.
+func negotiateContentType(r *http.Request) string {
+	contentTypes := []string{
+		contentTypePlainText,
+		contentTypeJSON,
+	}
+	return httputil.NegotiateContentType(r, contentTypes, contentTypePlainText)
+}
+
+// writeResponse is content-type aware response writer.
+func writeResponse(w http.ResponseWriter, r *http.Request, response generatedResponse) error {
+	if response.Err != "" {
+		w.WriteHeader(http.StatusInternalServerError)
+	} else {
+		w.WriteHeader(http.StatusOK)
+	}
+
+	switch negotiateContentType(r) {
```

Comment:
What if the header is unset, as it usually is?

Reply:
Then `negotiateContentType()` defaults to `text/plain` (and this is covered by unit test). So, whenever nothing is specified, we basically have the previous functionality, if "Accept:" header is provided, then we parse and use it.